File: 322.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<vector<int>> dp(coins.size() + 1, vector<int>(amount + 1, INT32_MAX));
        dp[0][0] = 0;

        for (int i = 1; i <= coins.size(); i++) {
            dp[i][0] = 0;
            for (int j = 1; j <= amount; j++) {
                int noTake = dp[i - 1][j];
                int take = INT32_MAX;
                if (j >= coins[i - 1] && dp[i][j - coins[i - 1]] != INT32_MAX) take = dp[i][j - coins[i - 1]] + 1;
                dp[i][j] = min(noTake, take);
            }
        }
        return (dp[coins.size()][amount] == INT32_MAX ? -1 : dp[coins.size()][amount]);
    }
};

int main() {
    Solution s;
    vector<int> coins = {2};
    int amount = 3;
    cout << s.coinChange(coins, amount) << endl;
    return 0;
}

File: 322.py
from ast import List
from cmath import inf

class Solution:
    def coinChange(self, coins, amount) -> int:
        dp = [[inf] * (amount + 1) for _ in range(len(coins) + 1)]
        for i in range(1, len(coins) + 1):
            for j in range(0, amount + 1):
                if j == 0:
                    dp[i][j] = 0
                    continue
                if j - coins[i - 1] >= 0 and dp[i][j - coins[i - 1]] != inf:
                    dp[i][j] = min(1 + dp[i][j - coins[i - 1]], dp[i - 1][j])
                else:
                    dp[i][j] = dp[i - 1][j]
        if dp[len(coins)][amount] != inf:
            return dp[len(coins)][amount]
        return -1

s = Solution
print(s.coinChange(s, coins=[2147483647], amount=2))